### PR TITLE
feature/particle cem model

### PR DESCRIPTION
**Pull request by PabloCorcho**
This PR includes a new Chemical Evolution Model based on a grid of particles that represent SSPs.

### CODE REVIEW

File: src/pst/models.py [hunk 1]
```diff
@@ -787,7 +789,118 @@ def integral_Z_SFR(self, time):
...
+
+    @masses.setter
+    def masses(self, values):
+        if not isinstance(values, u.Quantity):
+            self._masses = values << u.Msun
+        else:
+            self._masses = values
+
+    def interpolate_ssp_masses(self, ssp: SSPBase, t_obs: u.Quantity):
+        """Interpolate the particles to a base of SSPs.
```
Comment (paranoya) [inline src/pst/models.py:849]: Isn't there a method in `SSP` to do this?

File: src/pst/models.py [hunk 2]
```diff
@@ -787,7 +789,118 @@ def integral_Z_SFR(self, time):
...
+                    )
+
+        weights = np.zeros(ssp.L_lambda.shape[:-1]) << self.masses.unit
+        weights[metalliticy_idx, age_idx] = weight_Z * weight_age * self.masses[valid_particles]
+        weights[metalliticy_idx - 1, age_idx] = (1 - weight_Z) * weight_age * self.masses[valid_particles]
+        weights[metalliticy_idx - 1, age_idx - 1] = (1 - weight_Z) * (1 - weight_age) * self.masses[valid_particles]
+        weights[metalliticy_idx, age_idx - 1] = weight_Z * (1 - weight_age) * self.masses[valid_particles]
+
+        return weights
+
```
Comment (paranoya) [inline src/pst/models.py:890]: Especially if the galaxy contains a lot of particles, the tables `sort_idx`, `mass_history`, and `z_mass_history` probably be done once, upon initialisation

I'd actually argue for more informative names, such as e.g. `sorted_times`, `stellar_mass_formed`, and... err... `intergal_Z_SFR_table`? I cannot think of a sensible name, but it should be the same across all classes.

File: src/pst/models.py [hunk 3]
```diff
@@ -787,7 +789,118 @@ def integral_Z_SFR(self, time):
...
+    def integral_SFR(self, time):
+        sort_idx = np.argsort(self.time_form)
+        mass_history = np.cumsum(self.masses[sort_idx])
+        return np.interp(time, self.time_form[sort_idx], mass_history)
+
+    def integral_Z_SFR(self, time):
```
Comment (paranoya) [inline src/pst/models.py:896]: I may have changed my mind. I'd only store the sorted times (the time-consuming bit) and do the sums on the fly. More importantly, I'm not sure these functions make much sense for this class...

File: src/pst/models.py (hunk 1 above)
Comment (PabloCorcho) [reply src/pst/models.py:849]: Yes, but this one is in 2D

File: src/pst/models.py (hunk 2 above)
Comment (PabloCorcho) [reply src/pst/models.py:890]: What is the use of z_mass_history in this context?

File: src/pst/models.py (hunk 3 above)
Comment (PabloCorcho) [reply src/pst/models.py:896]: I agree, but according to out schema they are supposed to be abstractmethods. Maybe we should make the base class more flexible.

File: src/pst/models.py (hunk 1 above)
Comment (paranoya) [reply src/pst/models.py:849]: OK. As a matter of fact, we should probably remove `compute_SED` and `compute_burstSED` from `SSP` anyway.

File: src/pst/models.py (hunk 3 above)
Comment (paranoya) [reply src/pst/models.py:896]: Indeed. Or perhaps switch to `mass_weighted_stellar_metallicity`?

**Review by paranoya — APPROVED** [review]

**Pull request merged**